File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/formularios/FormFieldRow.kt
package br.church.paz.android.ui.features.formularios

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.DatePicker
import androidx.compose.material3.DatePickerDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import br.church.paz.android.ui.theme.PazShapes
import br.church.paz.android.ui.theme.PazSpacing
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Renders one form field (label + input) for a given [FormFieldDef]. Reused by both the
 * scrollable all-at-once [FormDetailScreen] and the one-question-per-screen [FormStepScreen].
 */
@Composable
fun FieldRow(
    def: FormFieldDef,
    value: String,
    isSubmitting: Boolean,
    uiState: FormDetailUiState,
    onValueChange: (String) -> Unit,
    onOpenPicker: (FormFieldDef) -> Unit,
    onSelfOrSearchMode: (String, Boolean) -> Unit,
    focusRequester: FocusRequester? = null,
    imeAction: ImeAction = ImeAction.Next,
    onImeAction: () -> Unit = {},
) {
    Column {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(PazSpacing.Xs),
        ) {
            Text(def.label, style = MaterialTheme.typography.labelMedium)
            if (def.required) {
                Text(
                    "*",
                    style = MaterialTheme.typography.labelMedium.copy(color = MaterialTheme.colorScheme.error),
                )
            }
        }
        Spacer(Modifier.height(PazSpacing.Sm))

        when (def.fieldType) {
            FormFieldType.PICKER ->
                PickerField(
                    value = value,
                    options = def.options,
                    enabled = !isSubmitting,
                    onValueChange = onValueChange,
                )

            FormFieldType.DATE -> DateFieldRow(value = value, enabled = !isSubmitting, onValueChange = onValueChange)

            FormFieldType.PHONE ->
                MaskedTextField(
                    fieldKey = def.key,
                    value = value,
                    placeholder = def.placeholder,
                    enabled = !isSubmitting,
                    keyboardType = KeyboardType.Number,
                    mask = { old, new -> applyPhoneMask(old = old, new = new) },
                    onValueChange = onValueChange,
                    focusRequester = focusRequester,
                    imeAction = imeAction,
                    onImeAction = onImeAction,
                )

            FormFieldType.EMAIL ->
                OutlinedTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth().focusable(focusRequester),
                    placeholder = { Text(def.placeholder) },
                    singleLine = true,
                    enabled = !isSubmitting,
                    keyboardOptions =
                        KeyboardOptions(
                            keyboardType = KeyboardType.Email,
                            capitalization = KeyboardCapitalization.None,
                            imeAction = imeAction,
                        ),
                    keyboardActions = androidx.compose.foundation.text.KeyboardActions(onNext = { onImeAction() }, onDone = { onImeAction() }),
                    shape = PazShapes.large,
                )

            FormFieldType.NAME ->
                OutlinedTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth().focusable(focusRequester),
                    placeholder = { Text(def.placeholder) },
                    singleLine = true,
                    enabled = !isSubmitting,
                    keyboardOptions =
                        KeyboardOptions(
                            keyboardType = KeyboardType.Text,
                            capitalization = KeyboardCapitalization.Words,
                            imeAction = imeAction,
                        ),
                    keyboardActions = androidx.compose.foundation.text.KeyboardActions(onNext = { onImeAction() }, onDone = { onImeAction() }),
                    shape = PazShapes.large,
                )

            FormFieldType.INTEGER ->
                OutlinedTextField(
                    value = value,
                    onValueChange = { new -> onValueChange(new.filter { it.isDigit() }) },
                    modifier = Modifier.fillMaxWidth().focusable(focusRequester),
                    placeholder = { Text(def.placeholder) },
                    singleLine = true,
                    enabled = !isSubmitting,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number, imeAction = imeAction),
                    keyboardActions = androidx.compose.foundation.text.KeyboardActions(onNext = { onImeAction() }, onDone = { onImeAction() }),
                    shape = PazShapes.large,
                )

            FormFieldType.CURRENCY ->
                MaskedTextField(
                    fieldKey = def.key,
                    value = value,
                    placeholder = def.placeholder,
                    enabled = !isSubmitting,
                    keyboardType = KeyboardType.Number,
                    mask = { _, new -> applyCurrencyMask(new) },
                    onValueChange = onValueChange,
                    focusRequester = focusRequester,
                    imeAction = imeAction,
                    onImeAction = onImeAction,
                )

            FormFieldType.MULTILINE ->
                OutlinedTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth().height(120.dp).focusable(focusRequester),
                    placeholder = { Text(def.placeholder) },
                    singleLine = false,
                    enabled = !isSubmitting,
                    keyboardOptions =
                        KeyboardOptions(
                            keyboardType = KeyboardType.Text,
                            capitalization = KeyboardCapitalization.Sentences,
                        ),
                    shape = PazShapes.large,
                )

            FormFieldType.BOOLEAN ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier.fillMaxWidth(),
                ) {
                    Switch(
                        checked = value == "true",
                        onCheckedChange = { onValueChange(if (it) "true" else "false") },
                        enabled = !isSubmitting,
                    )
                }

            FormFieldType.SELECT -> {
                val displayValue = if (def.optionValues.isEmpty()) {
                    value
                } else {
                    def.options.getOrElse(def.optionValues.indexOf(value)) { value }
                }
                PickerField(
                    value = displayValue,
                    options = def.options,
                    enabled = !isSubmitting,
                    onValueChange = { label ->
                        val apiValue = if (def.optionValues.isEmpty()) {
                            label
                        } else {
                            def.optionValues.getOrElse(def.options.indexOf(label)) { label }
                        }
                        onValueChange(apiValue)
                    },
                )
            }

            FormFieldType.USER_PICKER, FormFieldType.USER_MULTI_PICKER -> {
                val displayName = uiState.fields["${def.key}_name"] ?: ""
                Box {
                    OutlinedTextField(
                        value = displayName,
                        onValueChange = {},
                        modifier = Modifier.fillMaxWidth(),
                        readOnly = true,
                        enabled = !isSubmitting,
                        placeholder = { Text(def.placeholder.ifEmpty { "Selecionar pessoa" }) },
                        trailingIcon = {
                            Icon(Icons.Default.KeyboardArrowDown, null)
                        },
                        shape = PazShapes.large,
                        singleLine = true,
                    )
                    Surface(
                        modifier = Modifier.fillMaxWidth(),
                        onClick = { if (!isSubmitting) onOpenPicker(def) },
                        color = Color.Transparent,
                    ) {}
                }
            }

            FormFieldType.LG_PICKER -> {
                val displayName = uiState.fields["${def.key}_name"] ?: ""
                Box {
                    OutlinedTextField(
                        value = displayName,
                        onValueChange = {},
                        modifier = Modifier.fillMaxWidth(),
                        readOnly = true,
                        enabled = !isSubmitting,
                        placeholder = { Text(def.placeholder.ifEmpty { "Selecionar life group" }) },
                        trailingIcon = { Icon(Icons.Default.KeyboardArrowDown, null) },
                        shape = PazShapes.large,
                        singleLine = true,
                    )
                    Surface(
                        modifier = Modifier.fillMaxWidth(),
                        onClick = { if (!isSubmitting) onOpenPicker(def) },
                        color = Color.Transparent,
                    ) {}
                }
            }

            FormFieldType.SELF_OR_SEARCH -> {
                val isSearchMode = uiState.selfOrSearchIsSearch[def.key] == true
                Column {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        FilterChip(
                            selected = !isSearchMode,
                            onClick = { onSelfOrSearchMode(def.key, false) },
                            label = { Text("Eu mesmo") },
                        )
                        Spacer(Modifier.width(PazSpacing.Sm))
                        FilterChip(
                            selected = isSearchMode,
                            onClick = { onSelfOrSearchMode(def.key, true) },
                            label = { Text("Buscar pessoa") },
                        )
                    }
                    if (isSearchMode) {
                        Spacer(Modifier.height(PazSpacing.Sm))
                        val displayName = uiState.fields["${def.key}_name"] ?: ""
                        Box {
                            OutlinedTextField(
                                value = displayName,
                                onValueChange = {},
                                modifier = Modifier.fillMaxWidth(),
                                readOnly = true,
                                enabled = !isSubmitting,
                                placeholder = { Text("Selecionar pessoa") },
                                trailingIcon = { Icon(Icons.Default.KeyboardArrowDown, null) },
                                shape = PazShapes.large,
                            )
                            Surface(
                                modifier = Modifier.fillMaxWidth(),
                                onClick = { if (!isSubmitting) onOpenPicker(def) },
                                color = Color.Transparent,
                            ) {}
                        }
                    }
                }
            }

            FormFieldType.TEXT ->
                OutlinedTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth().focusable(focusRequester),
                    placeholder = { Text(def.placeholder) },
                    singleLine = true,
                    enabled = !isSubmitting,
                    keyboardOptions =
                        KeyboardOptions(
                            keyboardType = KeyboardType.Text,
                            capitalization = KeyboardCapitalization.Sentences,
                            imeAction = imeAction,
                        ),
                    keyboardActions = androidx.compose.foundation.text.KeyboardActions(onNext = { onImeAction() }, onDone = { onImeAction() }),
                    shape = PazShapes.large,
                )
        }
    }
}

private fun Modifier.focusable(focusRequester: FocusRequester?): Modifier =
    if (focusRequester != null) this.focusRequester(focusRequester) else this

// TextFieldValue gives us cursor control — without it, every mask application
// resets the cursor to position 0, making the field appear broken.
// [fieldKey] re-seeds the remembered TextFieldValue and cursor position whenever the
// underlying question changes, so a stable field instance (reused across step-mode
// screens) doesn't leak cursor/selection state between different questions.
@Composable
fun MaskedTextField(
    fieldKey: String,
    value: String,
    placeholder: String,
    enabled: Boolean,
    keyboardType: KeyboardType,
    mask: (old: String, new: String) -> String,
    onValueChange: (String) -> Unit,
    focusRequester: FocusRequester? = null,
    imeAction: ImeAction = ImeAction.Next,
    onImeAction: () -> Unit = {},
) {
    var tfv by remember(fieldKey) { mutableStateOf(TextFieldValue(value, selection = TextRange(value.length))) }

    LaunchedEffect(fieldKey) {
        tfv = TextFieldValue(value, selection = TextRange(value.length))
    }

    LaunchedEffect(value) {
        if (tfv.text != value) {
            tfv = TextFieldValue(value, selection = TextRange(value.length))
        }
    }

    OutlinedTextField(
        value = tfv,
        onValueChange = { incoming ->
            val masked = mask(tfv.text, incoming.text)
            tfv = TextFieldValue(masked, selection = TextRange(masked.length))
            onValueChange(masked)
        },
        modifier = Modifier.fillMaxWidth().focusable(focusRequester),
        placeholder = { Text(placeholder) },
        singleLine = true,
        enabled = enabled,
        keyboardOptions = KeyboardOptions(keyboardType = keyboardType, imeAction = imeAction),
        keyboardActions = androidx.compose.foundation.text.KeyboardActions(onNext = { onImeAction() }, onDone = { onImeAction() }),
        shape = PazShapes.large,
    )
}

@Composable
fun PickerField(
    value: String,
    options: List<String>,
    enabled: Boolean,
    onValueChange: (String) -> Unit,
) {
    var expanded by remember { mutableStateOf(false) }

    Box {
        OutlinedTextField(
            value = value,
            onValueChange = {},
            modifier = Modifier.fillMaxWidth(),
            readOnly = true,
            enabled = enabled,
            trailingIcon = {
                Icon(
                    if (expanded) Icons.Default.KeyboardArrowUp else Icons.Default.KeyboardArrowDown,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurface,
                )
            },
            shape = PazShapes.large,
            singleLine = true,
        )

        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false },
            modifier = Modifier.fillMaxWidth(0.9f),
        ) {
            options.forEach { option ->
                DropdownMenuItem(
                    text = { Text(option) },
                    onClick = {
                        onValueChange(option)
                        expanded = false
                    },
                    leadingIcon = if (value == option) {
                        {
                            Icon(
                                Icons.Filled.Check,
                                contentDescription = null,
                                tint = MaterialTheme.colorScheme.primary,
                            )
                        }
                    } else {
                        null
                    },
                )
            }
        }

        Surface(
            modifier = Modifier.fillMaxWidth(),
            onClick = { if (enabled) expanded = !expanded },
            color = Color.Transparent,
        ) {
            // Invisible surface to handle clicks
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DateFieldRow(
    value: String,
    enabled: Boolean,
    onValueChange: (String) -> Unit,
) {
    val fmt = remember { SimpleDateFormat("dd/MM/yyyy", Locale("pt", "BR")) }
    var showDialog by remember { mutableStateOf(false) }

    val initialMs =
        remember(value) {
            if (value.isNotEmpty()) fmt.parse(value)?.time else null
        } ?: System.currentTimeMillis()

    val pickerState = rememberDatePickerState(initialSelectedDateMillis = initialMs)

    OutlinedTextField(
        value = value,
        onValueChange = {},
        modifier = Modifier.fillMaxWidth(),
        placeholder = { Text("DD/MM/YYYY") },
        singleLine = true,
        enabled = enabled,
        readOnly = true,
        trailingIcon = {
            IconButton(onClick = { if (enabled) showDialog = true }) {
                Icon(Icons.Default.CalendarMonth, contentDescription = "Selecionar data")
            }
        },
        shape = PazShapes.large,
    )

    if (showDialog) {
        DatePickerDialog(
            onDismissRequest = { showDialog = false },
            confirmButton = {
                TextButton(onClick = {
                    pickerState.selectedDateMillis?.let { ms ->
                        onValueChange(fmt.format(Date(ms)))
                    }
                    showDialog = false
                }) { Text("OK") }
            },
            dismissButton = {
                TextButton(onClick = { showDialog = false }) { Text("Cancelar") }
            },
        ) {
            DatePicker(state = pickerState)
        }
    }
}

// Input helpers

fun applyPhoneMask(
    old: String,
    new: String,
): String {
    var digits = new.filter { it.isDigit() }
    val oldDigits = old.filter { it.isDigit() }
    // User deleted a separator character — drop the preceding digit too
    if (new.length < old.length && digits.length == oldDigits.length && digits.isNotEmpty()) {
        digits = digits.dropLast(1)
    }
    return formatPhone(digits.take(11))
}

// Separators go BEFORE the digit at boundary positions — no trailing chars at partial input
private fun formatPhone(digits: String): String {
    if (digits.isEmpty()) return ""
    val sb = StringBuilder()
    for ((i, c) in digits.withIndex()) {
        when (i) {
            0 -> sb.append("($c")
            1 -> sb.append("$c")
            2 -> sb.append(") $c") // ") " inserted before 3rd digit
            3 -> sb.append(" $c") // " " inserted before 4th digit
            7 -> sb.append("-$c") // "-" inserted before 8th digit
            else -> sb.append(c)
        }
    }
    return sb.toString()
}

fun applyCurrencyMask(input: String): String {
    val digits = input.filter { it.isDigit() }
    if (digits.isEmpty()) return ""
    val value = digits.toLongOrNull() ?: return ""
    val reais = value / 100
    val centavos = value % 100
    val reaisStr = if (reais == 0L) "0" else formatThousands(reais)
    return "$reaisStr,${centavos.toString().padStart(2, '0')}"
}

private fun formatThousands(n: Long): String {
    val s = n.toString()
    val sb = StringBuilder()
    for ((i, c) in s.reversed().withIndex()) {
        if (i > 0 && i % 3 == 0) sb.insert(0, '.')
        sb.insert(0, c)
    }
    return sb.toString()
}

File: kmp-mobile/ios/PazChurch/Features/Formularios/FormFieldRow.swift
import SwiftUI

/// Renders one form field (label + input) for a given `FormFieldDef`. Reused by both the
/// scrollable all-at-once `FormDetailView` and the one-question-per-screen `FormStepView`.
struct FieldRow: View {
    let def: FormFieldDef
    let value: String
    let extraFields: [String: String]
    let isSubmitting: Bool
    let selfOrSearchModes: [String: Bool]
    let onChange: (String) -> Void
    let onOpenPicker: (FormFieldDef) -> Void
    let onSelfOrSearchMode: (String, Bool) -> Void

    /// Only used by `FormStepView` — hooks the field's keyboard "next"/"return" action to
    /// advance to the next question, and applies the right submit label for the last step.
    var isFocused: FocusState<Bool>.Binding?
    var submitLabel: SubmitLabel = .next
    var onSubmitField: () -> Void = {}

    var body: some View {
        VStack(alignment: .leading, spacing: PazSpacing.sm) {
            HStack(spacing: 4) {
                Text(def.label)
                    .font(PazTypography.labelMedium)
                if def.required {
                    Text("*")
                        .font(PazTypography.labelMedium)
                        .foregroundColor(PazColors.error)
                }
            }

            switch def.fieldType {
            case .multiline:
                TextField(def.placeholder, text: Binding(get: { value }, set: onChange), axis: .vertical)
                    .font(PazTypography.bodyMedium)
                    .lineLimit(4...8)
                    .padding(PazSpacing.sm)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .disabled(isSubmitting)
                    .applyFocus(isFocused)
                    .submitLabel(submitLabel)
                    .onSubmit(onSubmitField)

            case .date:
                DateFieldRow(value: value, onChange: onChange, disabled: isSubmitting)

            case .phone:
                MaskedTextField(
                    placeholder: def.placeholder,
                    initialValue: value,
                    disabled: isSubmitting,
                    keyboardType: .numberPad,
                    contentType: .telephoneNumber,
                    mask: applyPhoneMask,
                    onChange: onChange,
                    isFocused: isFocused,
                    submitLabel: submitLabel,
                    onSubmitField: onSubmitField
                )

            case .email:
                TextField(def.placeholder, text: Binding(get: { value }, set: onChange))
                    .font(PazTypography.bodyMedium)
                    .keyboardType(.emailAddress)
                    .textContentType(.emailAddress)
                    .textInputAutocapitalization(.never)
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .disabled(isSubmitting)
                    .applyFocus(isFocused)
                    .submitLabel(submitLabel)
                    .onSubmit(onSubmitField)

            case .name:
                TextField(def.placeholder, text: Binding(get: { value }, set: onChange))
                    .font(PazTypography.bodyMedium)
                    .textContentType(.name)
                    .textInputAutocapitalization(.words)
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .disabled(isSubmitting)
                    .applyFocus(isFocused)
                    .submitLabel(submitLabel)
                    .onSubmit(onSubmitField)

            case .integer:
                TextField(def.placeholder, text: Binding(
                    get: { value },
                    set: { new in onChange(new.filter(\.isNumber)) }
                ))
                .font(PazTypography.bodyMedium)
                .keyboardType(.numberPad)
                .padding(.horizontal, PazSpacing.md)
                .frame(height: 56)
                .background(PazColors.surface)
                .clipShape(RoundedRectangle(cornerRadius: 12))
                .disabled(isSubmitting)
                .applyFocus(isFocused)
                .submitLabel(submitLabel)
                .onSubmit(onSubmitField)

            case .currency:
                MaskedTextField(
                    placeholder: def.placeholder,
                    initialValue: value,
                    disabled: isSubmitting,
                    keyboardType: .numberPad,
                    mask: { _, new in applyCurrencyMask(new) },
                    onChange: onChange,
                    isFocused: isFocused,
                    submitLabel: submitLabel,
                    onSubmitField: onSubmitField
                )

            case .toggle:
                Toggle(isOn: Binding(
                    get: { value == "true" },
                    set: { onChange($0 ? "true" : "false") }
                )) {
                    EmptyView()
                }
                .disabled(isSubmitting)

            case .text:
                TextField(def.placeholder, text: Binding(get: { value }, set: onChange))
                    .font(PazTypography.bodyMedium)
                    .textInputAutocapitalization(.sentences)
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .disabled(isSubmitting)
                    .applyFocus(isFocused)
                    .submitLabel(submitLabel)
                    .onSubmit(onSubmitField)

            case .select:
                let displayValue: String = {
                    if def.optionValues.isEmpty { return value }
                    guard let idx = def.optionValues.firstIndex(of: value) else { return value }
                    return def.options[idx]
                }()
                Menu {
                    ForEach(Array(def.options.enumerated()), id: \.offset) { idx, label in
                        Button(label) {
                            let apiValue = def.optionValues.isEmpty ? label : def.optionValues[idx]
                            onChange(apiValue)
                        }
                    }
                } label: {
                    HStack {
                        Text(displayValue
                            .isEmpty ? (def.placeholder.isEmpty ? "Selecionar" : def.placeholder) : displayValue)
                            .font(PazTypography.bodyMedium)
                            .foregroundStyle(displayValue.isEmpty ? PazColors.slate : PazColors.ink)
                        Spacer()
                        Image(systemName: "chevron.down").foregroundStyle(PazColors.accent)
                    }
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                }
                .disabled(isSubmitting)

            case .userPicker, .userMultiPicker:
                let displayName = extraFields["\(def.key)_name"] ?? ""
                Button(action: { if !isSubmitting { onOpenPicker(def) } }) {
                    HStack {
                        Text(displayName.isEmpty ? "Selecionar pessoa" : displayName)
                            .font(PazTypography.bodyMedium)
                            .foregroundStyle(displayName.isEmpty ? PazColors.slate : PazColors.ink)
                        Spacer()
                        Image(systemName: "chevron.down").foregroundStyle(PazColors.accent)
                    }
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                }
                .buttonStyle(.plain)

            case .lgPicker:
                let displayName = extraFields["\(def.key)_name"] ?? ""
                Button(action: { if !isSubmitting { onOpenPicker(def) } }) {
                    HStack {
                        Text(displayName.isEmpty ? "Selecionar life group" : displayName)
                            .font(PazTypography.bodyMedium)
                            .foregroundStyle(displayName.isEmpty ? PazColors.slate : PazColors.ink)
                        Spacer()
                        Image(systemName: "chevron.down").foregroundStyle(PazColors.accent)
                    }
                    .padding(.horizontal, PazSpacing.md)
                    .frame(height: 56)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                }
                .buttonStyle(.plain)

            case .selfOrSearch:
                let isSearchMode = selfOrSearchModes[def.key] == true
                VStack(alignment: .leading, spacing: PazSpacing.sm) {
                    HStack(spacing: PazSpacing.sm) {
                        Button("Eu mesmo") { onSelfOrSearchMode(def.key, false) }
                            .buttonStyle(.bordered)
                            .tint(isSearchMode ? .secondary : PazColors.accent)
                        Button("Buscar pessoa") { onSelfOrSearchMode(def.key, true) }
                            .buttonStyle(.bordered)
                            .tint(isSearchMode ? PazColors.accent : .secondary)
                    }
                    if isSearchMode {
                        let displayName = extraFields["\(def.key)_name"] ?? ""
                        Button(action: { if !isSubmitting { onOpenPicker(def) } }) {
                            HStack {
                                Text(displayName.isEmpty ? "Selecionar pessoa" : displayName)
                                    .font(PazTypography.bodyMedium)
                                    .foregroundStyle(displayName.isEmpty ? PazColors.slate : PazColors.ink)
                                Spacer()
                                Image(systemName: "chevron.down").foregroundStyle(PazColors.accent)
                            }
                            .padding(.horizontal, PazSpacing.md)
                            .frame(height: 56)
                            .background(PazColors.surface)
                            .clipShape(RoundedRectangle(cornerRadius: 12))
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
        }
    }
}

private extension View {
    /// Applies an optional `FocusState` binding — used only in step mode, where a single
    /// stable text field is reused across questions to keep the keyboard from flickering.
    @ViewBuilder
    func applyFocus(_ isFocused: FocusState<Bool>.Binding?) -> some View {
        if let isFocused {
            self.focused(isFocused)
        } else {
            self
        }
    }
}

/// Uses local @State so the mask runs inside onChange(of:) — the only reliable
/// way to intercept and replace text in SwiftUI without cursor/state conflicts.
struct MaskedTextField: View {
    let placeholder: String
    let initialValue: String
    let disabled: Bool
    var keyboardType: UIKeyboardType = .default
    var contentType: UITextContentType?
    let mask: (String, String) -> String // (old, new) -> masked
    let onChange: (String) -> Void
    var isFocused: FocusState<Bool>.Binding?
    var submitLabel: SubmitLabel = .next
    var onSubmitField: () -> Void = {}

    @State private var text: String = ""

    var body: some View {
        TextField(placeholder, text: $text)
            .font(PazTypography.bodyMedium)
            .keyboardType(keyboardType)
            .textContentType(contentType)
            .padding(.horizontal, PazSpacing.md)
            .frame(height: 56)
            .background(PazColors.surface)
            .clipShape(RoundedRectangle(cornerRadius: 12))
            .disabled(disabled)
            .applyFocus(isFocused)
            .submitLabel(submitLabel)
            .onSubmit(onSubmitField)
            .task { text = initialValue }
            .onChange(of: text) { old, new in
                let masked = mask(old, new)
                if masked != new { text = masked }
                onChange(masked)
            }
            .onChange(of: initialValue) { _, new in
                // Reseed when the underlying question changes (step mode reuses this
                // instance across different fields) — without this, stale text/cursor
                // state would leak between different questions of the same type.
                if new != text { text = new }
            }
    }
}

struct DateFieldRow: View {
    let value: String
    let onChange: (String) -> Void
    let disabled: Bool

    @State private var showPicker = false
    @State private var selected: Date = DateFormatter.brazilianDate
        .date(from: DateFormatter.brazilianDate.string(from: Date())) ?? Date()

    private let display = DateFormatter.brazilianDate

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Button(action: { if !disabled { showPicker.toggle() } }) {
                HStack {
                    Text(value.isEmpty ? "DD/MM/YYYY" : value)
                        .font(PazTypography.bodyMedium)
                        .foregroundStyle(value.isEmpty ? PazColors.slate : PazColors.ink)
                    Spacer()
                    Image(systemName: "calendar").foregroundStyle(PazColors.accent)
                }
                .padding(.horizontal, PazSpacing.md)
                .frame(height: 56)
                .background(PazColors.surface)
                .clipShape(RoundedRectangle(cornerRadius: 12))
            }
            .buttonStyle(.plain)
            .task {
                if let d = display.date(from: value) { selected = d }
                if value.isEmpty { onChange(display.string(from: Date())) }
            }

            if showPicker {
                DatePicker("", selection: $selected, displayedComponents: .date)
                    .datePickerStyle(.graphical)
                    .tint(PazColors.accent)
                    .onChange(of: selected) { _, d in
                        onChange(display.string(from: d))
                        showPicker = false
                    }
                    .padding(PazSpacing.sm)
                    .background(PazColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
            }
        }
    }
}

// MARK: - Input helpers

private func applyPhoneMask(old: String, new: String) -> String {
    var digits = new.filter(\.isNumber)
    let oldDigits = old.filter(\.isNumber)
    // User deleted a separator character — drop the preceding digit too
    if new.count < old.count, digits.count == oldDigits.count, !digits.isEmpty {
        digits = String(digits.dropLast())
    }
    return formatPhone(String(digits.prefix(11)))
}

/// Separators go BEFORE the digit at boundary positions — no trailing chars at partial input
private func formatPhone(_ digits: String) -> String {
    let d = Array(digits)
    guard !d.isEmpty else { return "" }
    var result = ""
    for (i, c) in d.enumerated() {
        switch i {
        case 0: result = "(\(c)"
        case 1: result += "\(c)"
        case 2: result += ") \(c)" // ") " inserted before 3rd digit
        case 3: result += " \(c)" // " " inserted before 4th digit
        case 7: result += "-\(c)" // "-" inserted before 8th digit
        default: result += "\(c)"
        }
    }
    return result
}

private func applyCurrencyMask(_ input: String) -> String {
    let digits = input.filter(\.isNumber)
    guard !digits.isEmpty else { return "" }
    let value = Int64(digits) ?? 0
    let reais = value / 100
    let centavos = value % 100
    let reaisStr = reais == 0 ? "0" : formatThousands(reais)
    return "\(reaisStr),\(String(format: "%02d", centavos))"
}

private func formatThousands(_ n: Int64) -> String {
    var result = ""
    let s = String(n)
    for (i, c) in s.reversed().enumerated() {
        if i > 0, i % 3 == 0 { result = "." + result }
        result = String(c) + result
    }
    return result
}

extension DateFormatter {
    static let brazilianDate: DateFormatter = {
        let f = DateFormatter(); f.dateFormat = "dd/MM/yyyy"; return f
    }()
}
